MDANSE/Src/MDANSE/Framework/QVectors/CircularLatticeQVectors.py

```diff
@@ -49,78 +55,53 @@ class CircularLatticeQVectors(LatticeQVectors):
-
-        qMax = (
-            self._configuration["shells"]["last"]
-            + 0.5 * self._configuration["width"]["value"]
-        )
-
-        uvMax = np.ceil([qMax / Vector(v).length() for v in qVects.T]) + 1
-        # Enforce integers in uvMax
-        uvMax = uvMax.astype(np.int64)
-
-        idxs = np.mgrid[-uvMax[0] : uvMax[0] + 1, -uvMax[1] : uvMax[1] + 1]
-
-        idxs = idxs.reshape(2, (2 * uvMax[0] + 1) * (2 * uvMax[1] + 1))
-
-        vects = np.dot(qVects, idxs)
-
-        dists2 = np.sum(vects**2, axis=0)
-
-        halfWidth = self._configuration["width"]["value"] / 2
-
-        nVectors = self._configuration["n_vectors"]["value"]
+        width = self._configuration["width"]["value"] / 2
 
         if self._status is not None:
             self._status.start(self._configuration["shells"]["number"])
 
         self._configuration["q_vectors"] = collections.OrderedDict()
 
         for q in self._configuration["shells"]["value"]:
-            qmin = max(0, q - halfWidth)
-
-            q2low = qmin * qmin
-            q2up = (q + halfWidth) * (q + halfWidth)
-
-            hits = np.where((dists2 >= q2low) & (dists2 <= q2up))[0]
-
-            nHits = len(hits)
-
-            if nHits != 0:
-                n = min(nHits, nVectors)
-
-                if nHits > nVectors:
-                    hits = random.sample(hits, nVectors)
-
-                self._configuration["q_vectors"][q] = {}
-                self._configuration["q_vectors"][q]["q_vectors"] = vects[:, hits]
-                self._configuration["q_vectors"][q]["n_q_vectors"] = n
-                self._configuration["q_vectors"][q]["q"] = q
-                self._configuration["q_vectors"][q]["hkls"] = self.qvectors_to_hkl(
-                    vects[:, hits], self._unit_cell
-                )
+            q_vectors = circle_of_vectors(q, width, nvecs_per_shell, rot_mat=rot_mat)
+            lattice_hkl_vectors, weights = self.lattice_vectors_with_weights(
+                q_vectors, self._unit_cell
+            )
+            selection = self.vectors_within_limits(
+                self.hkl_to_qvectors(lattice_hkl_vectors, self._unit_cell),
+                q_min=q - 0.5 * width,
+                q_max=q + 0.5 * width,
+            )
+            weights = weights[selection]
+            lattice_hkl_vectors = lattice_hkl_vectors.T[selection].T
+            if not len(weights):
+                self._configuration["q_vectors"][q] = None
+                continue
+
+            self._configuration["q_vectors"][q] = {}
+            self._configuration["q_vectors"][q]["q_vectors"] = self.hkl_to_qvectors(
+                lattice_hkl_vectors, self._unit_cell
+            )
+            self._configuration["q_vectors"][q]["weights"] = weights
+            self._configuration["q_vectors"][q]["n_q_vectors"] = np.sum(weights)
+            self._configuration["q_vectors"][q]["q"] = q
+            self._configuration["q_vectors"][q]["hkls"] = lattice_hkl_vectors
```

Comment:
Might be an opportunity to reduce repetition here.


```suggestion
            self._configuration["q_vectors"][q] = {
                "q_vectors": self.hkl_to_qvectors(lattice_hkl_vectors, self._unit_cell),
                "weights": weights,
                "n_q_vectors": np.sum(weights),
                "q": q,
                "hkls": lattice_hkl_vectors,
            }
```